File: engcrm-mobile/services/webLinks.ts
// Opening a stored website in the phone's browser.
//
// Website values reach the CRM from LLM enrichment, scraped directory
// listings, business-card OCR and hand typing, so they arrive without a
// scheme ("acme.de"), padded with whitespace, or carrying a scheme no browser
// should be handed. Linking.openURL rejects all of those with a thrown
// promise, so the screens go through here instead: normalise first, and let a
// bad value fail as plain unlinked text rather than an unhandled rejection.
//
// Parsing is done with a regex rather than `new URL()` because React Native's
// URL polyfill exposes no protocol/hostname getters.

import { Linking } from "react-native";

const HAS_SCHEME = /^[a-zA-Z][a-zA-Z0-9+.-]*:/;
const HTTP_URL = /^https?:\/\/([^/?#\s]+)/i;

/** Return an http(s) URL for `storedWebsite`, or null if it can't be one. */
export function browsableUrl(storedWebsite: string | null | undefined): string | null {
  const trimmed = (storedWebsite ?? "").trim();
  if (!trimmed) return null;
  const candidate = HAS_SCHEME.test(trimmed) ? trimmed : `https://${trimmed}`;
  const authority = HTTP_URL.exec(candidate)?.[1];
  const host = authority?.split("@").pop() ?? "";
  return host.includes(".") ? candidate : null;
}

/** Hand a stored website to the device's default browser. Never throws. */
export async function openWebsite(storedWebsite: string | null | undefined): Promise<boolean> {
  const url = browsableUrl(storedWebsite);
  if (!url) return false;
  try {
    await Linking.openURL(url);
    return true;
  } catch {
    return false;
  }
}

File: gcrm/api/web_links.py
"""Turn a stored website value into a URL a browser can safely be handed.

Website values reach the database from LLM enrichment, scraped directory
listings, business-card OCR and hand typing, so they arrive without a scheme
("acme.de"), padded with whitespace, or — worst case — as a "javascript:"
payload that would execute when someone clicks the rendered link. Every
template that renders a stored website as an anchor routes it through here.
"""
import re
from urllib.parse import urlparse

BROWSABLE_SCHEMES = ("http", "https")
_HAS_SCHEME = re.compile(r"^[a-zA-Z][a-zA-Z0-9+.\-]*:")


def browsable_url(stored_website: str | None) -> str | None:
    """Return an http(s) URL for `stored_website`, or None if it can't be one.

    A scheme-less value is assumed to be https. Anything carrying another
    scheme (javascript:, data:, mailto:) or lacking a host is rejected rather
    than linked, so a bad row degrades to plain text instead of a live link.
    """
    candidate = (stored_website or "").strip()
    if not candidate:
        return None
    if not _HAS_SCHEME.match(candidate):
        candidate = f"https://{candidate}"
    parsed = urlparse(candidate)
    if parsed.scheme not in BROWSABLE_SCHEMES or "." not in parsed.netloc:
        return None
    return candidate
